File: sum.py
// Time Complexity : O(N*N)
// Space Complexity :O(N*N) 
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> List[List[int]]:
        paths = []
        
        def helper(node, total, path):
            if node:
                if not node.left and not node.right:
                    if total + node.val == sum:
                        paths.append(path+[node.val])
                    return
                helper(node.left, total + node.val, path + [node.val])
                helper(node.right, total + node.val, path + [node.val])
                
        helper(root, 0, [])
        return paths

File: sum.java
// Time Complexity :O(N)
// Space Complexity :O(H)
// Did this code successfully run on Leetcode :	Yes
// Any problem you faced while coding this : No

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    List<List<Integer>> output = new ArrayList();
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        paths(root,new ArrayList(),sum);
        return output;
    }
    
    private void paths(TreeNode root,ArrayList list, int sum){

        if(root == null)	// base case
            return;

        list.add(root.val);		// add the root element to the list before checking for children element
        
        sum -=root.val;			// everytime we see a node, sum the node value with sum and recursively traverse until the leaf node is 0
        
        if(root.right == null && root.left ==null && sum ==0)	// if we reach the leaf node and sum is 0, add the list of all the root values to the list
            output.add(new ArrayList(list));

        paths(root.left,list,sum);		// traverse through the tree recursively
        paths(root.right,list,sum);
        list.remove(list.size()-1);
    }
}